Add logic for reflecting IDREF attributes

mrego opened the pull request:
This defines IDL reflection for single and FrozenArray attributes.
This work was done mostly by @alice.

This is just a rebase of the following PR: https://github.com/whatwg/html/pull/3917

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

- [X] At least two implementers are interested (and none opposed):
   * Chromium has an implementation under `AOMAriaRelationshipProperties` flag.
   * WebKit has an implementation under `AriaReflectionForElementReferencesEnabled` flag.
- [X] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/aria-element-reflection.tentative.html
- [X] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=981423
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1769586
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=196843


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/7934/canvas.html" title="Last updated on Jun 9, 2022, 8:42 PM UTC (edccf21)">/canvas.html</a>  ( <a href="https://whatpr.org/html/7934/2f901a7...edccf21/canvas.html" title="Last updated on Jun 9, 2022, 8:42 PM UTC (edccf21)">diff</a> )
<a href="https://whatpr.org/html/7934/common-dom-interfaces.html" title="Last updated on Jun 9, 2022, 8:42 PM UTC (edccf21)">/common-dom-interfaces.html</a>  ( <a href="https://whatpr.org/html/7934/2f901a7...edccf21/common-dom-interfaces.html" title="Last updated on Jun 9, 2022, 8:42 PM UTC (edccf21)">diff</a> )
<a href="https://whatpr.org/html/7934/infrastructure.html" title="Last updated on Jun 9, 2022, 8:42 PM UTC (edccf21)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/7934/2f901a7...edccf21/infrastructure.html" title="Last updated on Jun 9, 2022, 8:42 PM UTC (edccf21)">diff</a> )

domenic commented on source:
```diff
@@ -7566,8 +7568,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+  <li><p>Elements of the type this IDL attribute appears on have <dfn>cached
+  <var>attr</var>-associated elements</dfn>, which is a <code
+  data-x="">FrozenArray&lt;<code>Element</code>&gt;?</code>. It is initially null.</p></li>
```
Should this be T instead of Element?

mrego replied:
Done.

domenic commented on source:
```diff
@@ -7566,8 +7568,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+        <li><p>If <var>attrElement</var> is not a <span>descendant</span> of any of <var>element</var>'s <span
+        data-x="concept-shadow-including-ancestor">shadow-including ancestors</span>, then
+        <span>continue</span>.</p></li>
+        <li><p>Add <var>attrElement</var> to <var>elements</var>.</p></li>
```
Nit: use `<span data-x="list append">Append</span>` instead of `Add`

mrego replied:
Done.

domenic commented on source:
```diff
@@ -7566,8 +7568,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+  </li>
+
+  <li>
+    <p class="note">Other parts of this specification, or other specifications using attribute
```
This should be inside the previous `<li>` instead of in its own `<li>`

domenic commented on source:
```diff
@@ -7566,8 +7568,274 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+   <li>
+    <p>The following <span data-x="concept-element-attributes-change-ext">attribute change
+    steps</span> are used to synchronize between the content attribute and the IDL attribute:</p>
```
Hmm, I find the sudden introduction of _localName_, _namespace_, and _element_ a bit weird here. I think the best style would be to say "The following attribute change steps, given _element_, _localName_, _oldValue_, _value_, and _namespace_, are used...". Even though oldValue and value are unused.

Same for the other occurrence.

mrego replied:
Done.

domenic reviewed (commented):
This overall looks great. I'm excited to see implementations have happened since I last was reviewing this. I found a few editorial nits but I'd be happy to add this to the spec.

I am not really familiar with the shadow tree details so if you, or another reviewer, wants to confirm that the tests you linked match the spec draft here, that would be helpful.

I don't see tests for the caching invariant, e.g. testing that `assert_equals(input1.ariaLabelledByElements, input1.ariaLabelledByElements)` (in the array case, not just the trivial `null` case). Could you be sure those are added?

I will note that since this was originally written, `ObservableArray<T>` has become a thing, and using it instead of `FrozenArray<T>` would probably be more friendly to web developers. And I suspect that if we merge this now people won't necessarily put in the work to upgrade it over time, which is sad. But I don't want to stand in the way of people unflagging a feature with two implementations... Let us know if you have any thoughts on that.

mrego commented:
Thanks for the quick review. I've applied the suggested changes.

> I am not really familiar with the shadow tree details so if you, or another reviewer, wants to confirm that the tests you linked match the spec draft here, that would be helpful.

I'm not a Shadow DOM expert, but the tests seem to cover what this PR mentions. Maybe others could confirm that.

> I don't see tests for the caching invariant, e.g. testing that `assert_equals(input1.ariaLabelledByElements, input1.ariaLabelledByElements)` (in the array case, not just the trivial `null` case). Could you be sure those are added?

I'm adding them in https://github.com/web-platform-tests/wpt/pull/34095.

> I will note that since this was originally written, `ObservableArray<T>` has become a thing, and using it instead of `FrozenArray<T>` would probably be more friendly to web developers. And I suspect that if we merge this now people won't necessarily put in the work to upgrade it over time, which is sad. But I don't want to stand in the way of people unflagging a feature with two implementations... Let us know if you have any thoughts on that.

Looks like [WebKit doesn't have `ObservableArray` yet](https://bugs.webkit.org/show_bug.cgi?id=238281), so that might be a reason why this is still using `FrozenArray`. Not really sure though.
This has also been discussed in some AOM meetings in the past, but I don't know what was the final reason to go with `FrozenArray`
Maybe @rniwa has some thoughts on this.

mrego commented:
> > I don't see tests for the caching invariant, e.g. testing that `assert_equals(input1.ariaLabelledByElements, input1.ariaLabelledByElements)` (in the array case, not just the trivial `null` case). Could you be sure those are added?
> 
> I'm adding them in [web-platform-tests/wpt#34095](https://github.com/web-platform-tests/wpt/pull/34095).

On top of that I've added more tests to check this in different situations at https://github.com/web-platform-tests/wpt/pull/34194

And I've also fixed the WebKit implementation so now all the tests pass on WebKit, and the caching invariant works as expected with `FrozenArray`: https://bugs.webkit.org/show_bug.cgi?id=240563

In addition, Chromium has a TODO to fix this in the code (see [this comment](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.cc;l=999;drc=115b58435c649c4a3a7b7f29d4ba795e04afc1a3)).

Giving the above information, is this PR blocked on something else?

annevk commented:
I think we need to point out that an element's explicitly set _attr_-elements are weak references, no? The idea being that if the elements they point to can be collected, this shouldn't hold them alive as the corresponding _attr_-associated element will already have started returning null.

That also makes it clearer you really cannot depend on explicitly set _attr_-element. Perhaps we should stress that even more though by prefixing it with "internal" or some such?

I'm concerned as in https://github.com/WICG/aom/issues/192#issuecomment-1145675484 there's already a suggestion other things could build on top of these internal details, which I don't think is workable.

cc @smaug----

domenic commented:
Great work @mrego!

+1 to @annevk's concerns, especially making sure we're all on the same page that the explicitly set attr-element is an internal implementation detail and not something any other specs or parts of the implementation (including the accessibility tree) should build on.

mrego commented:
> I think we need to point out that an element's explicitly set _attr_-elements are weak references, no? The idea being that if the elements they point to can be collected, this shouldn't hold them alive as the corresponding _attr_-associated element will already have started returning null.

Good idea, also both implementations use weak references. I've uploaded a new commit adding some text about that.

> That also makes it clearer you really cannot depend on explicitly set _attr_-element. Perhaps we should stress that even more though by prefixing it with "internal" or some such?
> 
> I'm concerned as in [WICG/aom#192 (comment)](https://github.com/WICG/aom/issues/192#issuecomment-1145675484) there's already a suggestion other things could build on top of these internal details, which I don't think is workable.

I've already mentioned there that the idea should be discarded, it was just an idea on the last AOM meeting and people were not very aware of the this spec text.

I have the feeling that the spec text is pretty clear with the 2 notes:
```
    <p class="note">Other parts of this specification, or other specifications using attribute
    reflection, are generally expected to consult the <span><var>attr</var>-associated
    element</span>. The <span>explicitly set <var>attr</var>-element</span> is an internal
    implementation detail of the <span><var>attr</var>-associated element</span>, and is not to be
    used directly.</p>

   <p class="note">Other parts of this specification, or other specifications using attribute
   reflection, are generally expected to consult the <span><var>attr</var>-associated
   elements</span>. The <span>explicitly set <var>attr</var>-elements</span> are an internal
   implementation detail of the <span><var>attr</var>-associated elements</span>, and are not to be
   used directly. Similarly, the <span>cached <var>attr</var>-associated elements</span> are an
   internal implementation detail of the IDL attribute's getter.</p>
```

TBH, I'm not sure if naming them internal would have encouraged not to think on the possibility that was discussed in https://github.com/WICG/aom/issues/192#issuecomment-1145675484. Anyway if you think that it's going to help I guess we can just add "internal" to both the explicitly set and the cached one.

WDYT?

annevk commented on source:
```diff
@@ -7566,8 +7568,277 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   null.</p></li>
+
+   <li>
+    <p>Elements of the type this IDL attribute appears on have an <dfn data-export=""><var>attr</var>-associated
```
Nit: `<dfn for="Element" export>` (here and below)

mrego replied:
Done.

Let me know if I should do the same in other things too like `<dfn data-export=""><var>attr</var>-associated element</dfn>`, `<dfn>cached <var>attr</var>-associated elements</dfn>` or `<dfn data-export=""><var>attr</var>-associated elements</dfn>`.

annevk replied:
That's what I was commenting on. We shouldn't export the internal details :-)

mrego replied:
Ok, sorry for the confusion. I think I got it right now.

annevk commented on source:
```diff
@@ -7566,8 +7568,277 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    </ol>
+
+    <p class="note">Other parts of this specification, or other specifications using attribute
+    reflection, are generally expected to consult the <span><var>attr</var>-associated
```
Nit: let's strike generally (here and below).

mrego replied:
Done.

annevk commented on source:
```diff
@@ -7566,8 +7568,277 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    element</span>. The <span>explicitly set <var>attr</var>-element</span> is an internal
+    implementation detail of the <span><var>attr</var>-associated element</span>, and is not to be
+    used directly.</p>
+   </li>
```
For both notes I also think it would be clearer if we talked about "An element's _nameOfFieldHere_" to make it clear it's not a standalone concept. (This would also get rid of the pluralization in the second note, which looks a bit funny to me.)

mrego replied:
Good suggestion, changed that.

annevk reviewed (commented):
Thanks @mrego! I think you're correct that the notes should suffice. I have a number of tiny nits, but overall this looks great.

mrego commented on source:
```diff
@@ -7566,8 +7568,277 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    <p class="note">Other parts of this specification, or other specifications using attribute
+    reflection, are generally expected to consult the <span><var>attr</var>-associated
+    element</span>. The <span>explicitly set <var>attr</var>-element</span> is an internal
+    implementation detail of the <span><var>attr</var>-associated element</span>, and is not to be
```
Done.

annevk commented on source:
```diff
@@ -7566,8 +7568,277 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+     <li><p>Return null.</p></li>
+    </ol>
+
+    <p class="note">Other parts of this specification, or other specifications using attribute
+    reflection, are expected to consult the <span><var>attr</var>-associated element</span>. An
+    element's <span>explicitly set <var>attr</var>-element</span> is an internal implementation
+    detail of the <span><var>attr</var>-associated element</span> and is not to be used
+    directly.</p>
```
This is what I was thinking of:
```suggestion
    <p class="note">Other parts of this specification, or other specifications using attribute
    reflection, are expected to consult an element's <span><var>attr</var>-associated 
    element</span>. An element's <span>explicitly set <var>attr</var>-element</span> is an internal 
    implementation detail of its <span><var>attr</var>-associated element</span> and is not to be 
    used directly.</p>
```

domenic commented on source:
```diff
@@ -7566,8 +7568,277 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    <ol>
+     <li><p>Let <var>id</var> be the empty string.</p></li>
+
+     <li><p>If the given value:</p>
```
I think we need to handle the case of the given value being null here, before this step.

mrego replied:
Right, I've added a new step to handle that case.

domenic commented on source:
```diff
@@ -7566,8 +7568,277 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+   </li>
+
+   <li><p>On getting, the IDL attribute must return this element's <span><var>attr</var>-associated
+   element</span>.</p></li>
```
For both this and the FrozenArray version, can you change to "On getting/setter, the IDL attribute must" to "The getter/setters steps are" similar to what we recently introduced for DOMString/DOMString? in https://html.spec.whatwg.org/#reflecting-content-attributes-in-idl-attributes

mrego replied:
Done.

domenic reviewed (commented):
Only one substantive comment, which is that I believe we forgot the null handling for the `T?` case. Everything else is minor style updates (which I would have fixed myself before landing if it weren't for the more substantive issue.)

mrego commented on source:
```diff
@@ -7566,8 +7568,277 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+     <li><p>Set the content attribute's value for this element to <var>id</var>.</p></li>
 
+     <li><p>Set this element's <span>explicitly set <var>attr</var>-element</span> to be a weak
```
Done.

mrego commented on source:
```diff
@@ -7566,8 +7568,277 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+      <ol>
+       <li><p>If <var>value</var> is empty and <var>elements</var> is non-empty, then <span
+       data-x="list append">append</span> a weak reference to <var>element</var> to
+       <var>elements</var> and <span>continue</span>.</p></li>
```
I've tried to do this, but not totally sure if that was what you were looking for, please check.

domenic reviewed (approved):
LGTM! Will merge tomorrow.
